### PR TITLE
Faster semicolon finding and hashing

### PR DESCRIPTION
**Pull request by mjs**
Use bit masking to find the semicolon. Only hash the first 8 bytes (as a 64 bit int instead of a byte slice).

Inspiration taken from https://github.com/gunnarmorling/1brc/blob/main/src/main/java/dev/morling/onebrc/CalculateAverage_mtopolnik.java

### CODE REVIEW

**Comment by benhoyt** [comment]
Ooh, this is neat, thanks Menno. I haven't totally figured it out yet, but the results are good:

```
$ ./go-1brc -revision=1 ../1brc/data/measurements.txt >out-r1
Processed 13156.2MB in 1m39.507011009s
$ ./go-1brc -revision=9 ../1brc/data/measurements.txt >out-r9
Processed 13156.2MB in 2.893693843s  # 34.4x as fast as the r1 above
$ ./go-1brc -revision=10 ../1brc/data/measurements.txt >out-r10
Processed 13156.2MB in 2.497241029s  # 39.8x as fast as the r1 above
```

Not sure why the fast solutions are relatively faster now than they were when I tested the other week, but I'll take it! In any case, this r10 solution is consistently noticeably faster than r9.

**Comment by benhoyt** [comment]
It's quite neat how this can all still be done without unsafe with `NativeEndian` and the `bits` package.

I notice that Java solution also does some funky stuff to parse the temperature without loops:

```java
        private static int parseTemperature(long word, int dotPos) {
            final long signed = (~word << 59) >> 63;
            final long removeSignMask = ~(signed & 0xFF);
            final long digits = ((word & removeSignMask) << (28 - dotPos)) & 0x0F000F0F00L;
            final long absValue = ((digits * 0x640a0001) >>> 32) & 0x3FF;
            return (int) ((absValue ^ signed) - signed);
        }
```

I guess that'll be coming next. :-)

File: r10.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
+// r10: all the previous optimizations plus faster semicolon finding and
+// hashing
+//
+// 24.8x faster than r1
+// 47.62ms for 10M rows on Menno's laptop (which is slower than Ben's)
```
Comment (benhoyt) [inline r10.go]: Would you mind also leaving a link to https://github.com/gunnarmorling/1brc/blob/main/src/main/java/dev/morling/onebrc/CalculateAverage_mtopolnik.java in a code comment?

**Review by benhoyt — APPROVED** [review]

File: r10.go (hunk 1 above)
Comment (mjs) [reply r10.go]: Thanks for taking care of this. We've had guests staying with us so I haven't had a chance to get back to this.

**Comment by mjs** [comment]
> It's quite neat how this can all still be done without unsafe with `NativeEndian` and the `bits` package.
> 
> I notice that Java solution also does some funky stuff to parse the temperature without loops:
> 
> ```java
>         private static int parseTemperature(long word, int dotPos) {
>             final long signed = (~word << 59) >> 63;
>             final long removeSignMask = ~(signed & 0xFF);
>             final long digits = ((word & removeSignMask) << (28 - dotPos)) & 0x0F000F0F00L;
>             final long absValue = ((digits * 0x640a0001) >>> 32) & 0x3FF;
>             return (int) ((absValue ^ signed) - signed);
>         }
> ```
> 
> I guess that'll be coming next. :-)

That's certainly fair game for a future revision. This is very clever - no branches!

I also have a plan to reduce the amount of bytes.Equal calls when updating the per-worker maps. These currently dominate the CPU profile.

**Pull request merged**